src/coreclr/src/jit/optimizer.cpp

```diff
@@ -9189,3 +9189,133 @@ void Compiler::optOptimizeBools()
+         block             = block->GetUniqueSucc())
+    {
+        block->bbFlags |= BBF_MARKED;
+        for (Statement* stmt = block->FirstNonPhiDef(); stmt != nullptr;)
+        {
+            Statement* next = stmt->GetNextStmt();
+            for (GenTree* tree = stmt->GetTreeList(); tree != nullptr; tree = tree->gtNext)
+            {
+                if ((tree->gtFlags & GTF_CALL) != 0)
+                {
+                    hasGCSafePoint = true;
+                }
+
+                switch (tree->gtOper)
+                {
+                    case GT_LCL_VAR:
+                    case GT_LCL_FLD:
+                    case GT_LCL_VAR_ADDR:
+                    case GT_LCL_FLD_ADDR:
+                    {
+                        unsigned  lclNum    = tree->AsLclVarCommon()->GetLclNum();
+                        unsigned* pRefCount = refCounts.LookupPointer(lclNum);
+                        if (pRefCount != nullptr)
+                        {
+                            *pRefCount = (*pRefCount)++;
+                        }
+                        else
+                        {
+                            refCounts.Set(lclNum, 1);
+                        }
+
+                        break;
+                    }
+                    case GT_ASG:
+                    {
+                        GenTreeOp* treeOp = tree->AsOp();
+                        if (treeOp->gtOp1->OperIs(GT_LCL_VAR, GT_LCL_FLD))
+                        {
+                            unsigned         lclNum    = treeOp->gtOp1->AsLclVarCommon()->GetLclNum();
+                            LclVarDsc* const lclDsc    = lvaGetDesc(lclNum);
+                            unsigned*        pRefCount = refCounts.LookupPointer(lclNum);
+                            assert(pRefCount != nullptr);
+                            if (*pRefCount == 1)
+                            {
+                                // The local hasn't been referenced before this assignment.
+                                bool removedExplicitZeroInit = false;
+                                if (!lclDsc->lvTracked && treeOp->gtOp2->IsIntegralConst() &&
+                                    (treeOp->gtOp2->AsIntCon()->IconValue() == 0))
+                                {
+                                    bool bbInALoop  = (block->bbFlags & BBF_BACKWARD_JUMP) != 0;
+                                    bool bbIsReturn = block->bbJumpKind == BBJ_RETURN;
+
+                                    if (!fgVarNeedsExplicitZeroInit(lclDsc, bbInALoop, bbIsReturn))
+                                    {
+                                        // We are guaranteed to have a zero initialization in the prolog and
+                                        // the local hasn't been redefined between the prolog and this explicit
+                                        // zero initialization so the assignment can be safely removed.
+                                        if (tree == stmt->GetRootNode())
```

Comment:
For non-root assignments can't we just bash the asg tree to a nop? Or if bashing during walking is too painful, keep track of these trees and bash them later?

Or are these rare enough that it's not worth trying to handle them?

Reply:
I tried that and saw no new diffs in framework and benchmarks.